support using gradients named for outputs in derivatives

dagitses opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63947
* #64439

Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

Commit:
support using gradients named for outputs in derivatives

Fixes #62196

[ghstack-poisoned]

dagitses commented:
#63040 was where the conversation of this PR started. That PR was closed when this needed to become a stack.

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -67,6 +71,11 @@
 #
+#     Across a differentiable function's derivatives set, it is not
+#     permitted to mix the use of "grad", "grads[INDEX]", and
+#     "grad_NAME". You must be consistent for that differentiable
```
nit: "grad_{name}" ? To be consistent with the other sentence above.

dagitses replied:
done

albanD commented on tools/autograd/load_derivatives.py:
```diff
@@ -88,6 +91,9 @@ def create_derivative(f: NativeFunction, formula: str, var_names: Tuple[str, ...
     formula, saved_inputs = saved_variables(formula, arguments, var_names)
     formula, saved_outputs = saved_variables(formula, named_returns, var_names)
 
+    named_gradients = {name for name in available_named_gradients
```
nit: used_named_gradients ?

dagitses replied:
done. do you want me to change the field name on Derivative as well?

albanD commented on tools/autograd/load_derivatives.py:
```diff
@@ -303,12 +310,14 @@ def check_grad_usage(defn_name: str, derivatives: Sequence[Derivative]) -> None:
 
         used_grad = 0
         used_grads = 0
+        used_named_grad = False
```
This name is confusing. `uses_named_grad` or `using_named_grad` would be better.
And I would also change the two other above for consistency.

dagitses replied:
#64439 addresses the other two. and changed to `uses_named_grads` in this PR.

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

albanD commented on tools/autograd/load_derivatives.py:
```diff
@@ -422,7 +461,13 @@ def set_up_derivatives(f: NativeFunction) -> Tuple[
                            "Please use a different name in native_functions.yaml.")
 
-    derivatives, forward_derivatives, args_with_derivatives, non_differentiable_arg_names = set_up_derivatives(canonical)
+    (derivatives, forward_derivatives, args_with_derivatives,
```
This line splitting is pretty hard to read...

dagitses replied:
what do you think about this instead? this way we have a left side and a right side of the operator, despite it all spanning multiple lines.

dagitses replied:
test

albanD commented on tools/autograd/load_derivatives.py:
```diff
@@ -374,7 +412,8 @@ def set_up_derivatives(f: NativeFunction) -> Tuple[
         check_grad_usage(defn_name, derivatives)
 
-        return derivatives, forward_derivatives, args_with_derivatives, non_differentiable_arg_names
+        return (derivatives, forward_derivatives, args_with_derivatives,
```
Why do you need to return this?

dagitses replied:
In gen_autograd_functions.py we need to associate each name with its index so that we can add each `const auto grad_name = grads[index];` alias.

Per our earlier discussion in the PR before ghstack (#63040) we decided not to have a map of each used name and its index in each derivative. But we still need that mapping in the codegen so we return it.

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -9524,6 +9532,128 @@ def fn(x1, x2):
+                functions_by_schema={specification: native_function},
+            )
+
+class TestGenAutogradFunctions(TestCase):
```
Why do you use two classes here? One for all the codegen stuff should be enough?

dagitses replied:
Note: moved to its own file now.

I was thinking one test case per module. No good?

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).

albanD commented on tools/test/test_codegen.py:
```diff
@@ -0,0 +1,137 @@
+
+# Represents the most basic NativeFunction. Use dataclasses.replace()
+# to edit for use.
+DEFAULT_NATIVE_FUNCTION, _ = tools.codegen.model.NativeFunction.from_yaml(
```
nit why not at the top of the file?

albanD replied:
That sounds ok.

albanD reviewed (approved):
SGTM

Commit:
Update on "support using gradients named for outputs in derivatives"


Fixes #62196

Differential Revision: [D30541485](https://our.internmc.facebook.com/intern/diff/D30541485)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541485).
